commit/internal/builder/report.go

```diff
@@ -0,0 +1,396 @@
+			0,   // no RMN for partial reports.
+			priceUpdates,
+		)
+		if err != nil {
+			return nil, err
+		}
+		if report != nil {
+			reports = append(reports, *report)
+		}
+	}
+
+	return reports, nil
+}
+
+// buildMultiplePriceReports is a helper to splits up price data into multiple reports.
+// Helper for buildMultiplePriceAndMerkleRootReports.
+// Merkle root data is ignored.
+func buildMultiplePriceReports(
+	ctx context.Context,
+	lggr logger.Logger,
+	reportCodec cciptypes.CommitPluginCodec,
+	transmissionSchedule *ocr3types.TransmissionSchedule,
+	outcome committypes.Outcome,
+	maxPricesPerReport uint64, // passed in directly to avoid implementing ReportBuilderFunc
+) ([]ocr3types.ReportPlus[[]byte], error) {
+	// update joins together the different types of price updates so that they can be
+	// selected in one loop.
+	type update struct {
+		cciptypes.TokenPrice
+		cciptypes.GasPriceChain
+	}
+	// Helper to add an update to the PriceUpdates type.
+	add := func(updates cciptypes.PriceUpdates, update update) cciptypes.PriceUpdates {
+		if (update.TokenPrice != cciptypes.TokenPrice{}) {
+			updates.TokenPriceUpdates = append(updates.TokenPriceUpdates, update.TokenPrice)
+		}
+		if (update.GasPriceChain != cciptypes.GasPriceChain{}) {
+			updates.GasPriceUpdates = append(updates.GasPriceUpdates, update.GasPriceChain)
+		}
+		return updates
+	}
+
+	var updates []update
+	for _, tokenPriceUpdate := range outcome.TokenPriceOutcome.TokenPrices.ToSortedSlice() {
+		updates = append(updates, update{
+			TokenPrice: tokenPriceUpdate,
+		})
+	}
+	for _, gasPriceUpdate := range outcome.ChainFeeOutcome.GasPrices {
+		updates = append(updates, update{
+			GasPriceChain: gasPriceUpdate,
+		})
+	}
+
+	var reports []ocr3types.ReportPlus[[]byte]
+	numUpdates := uint64(0)
+	priceUpdates := cciptypes.PriceUpdates{}
+
+	for _, u := range updates {
+		numUpdates++
+		priceUpdates = add(priceUpdates, u)
+
+		if numUpdates == maxPricesPerReport {
+			report, err := buildOneReport(
+				ctx,
+				lggr,
+				reportCodec,
+				transmissionSchedule,
+				outcome.MerkleRootOutcome.OutcomeType,
+				nil,
+				nil,
+				nil, // no RMN for partial reports.
+				0,   // no RMN for partial reports.
+				priceUpdates,
+			)
+			if err != nil {
+				return nil, err
+			}
+			if report != nil {
+				reports = append(reports, *report)
+			}
+
+			// reset accumulators for next report.
+			numUpdates = 0
+
+			// price updates are only included in the first report.
+			priceUpdates = cciptypes.PriceUpdates{}
+		}
+	}
+
+	// check for final partial report.
+	if numUpdates > 0 {
+		report, err := buildOneReport(
+			ctx,
+			lggr,
+			reportCodec,
+			transmissionSchedule,
+			outcome.MerkleRootOutcome.OutcomeType,
+			nil,
+			nil,
+			nil, // no RMN for partial reports.
+			0,   // no RMN for partial reports.
+			priceUpdates,
+		)
+		if err != nil {
+			return nil, err
+		}
+		if report != nil {
+			reports = append(reports, *report)
+		}
```

Comment:
Ditto here maybe a check that `report == nil` and return an error since its an invariant violation (postcondition of buildOneReport should be that report is non-nil if price reports are provided, which in this case they are)

Reply:
I don't think we want to return an error. If other reports were generated they should still be submitted.

Reply:
It looks like this behavior is a carry-over from when this code was called from the OCR `Reports` function.

`buildOneReport` has special handling to check for `outcomeType == EmptyReport` and a check for `report.IsEmpty()`. If you want no-op reports to be an error we should change this behavior so that `buildOneReport` returns an error and we avoid calling it when there's no need.

Personally, I don't mind the condition-less logic:
1) pass all available data to the build function.
     a) if there's no data, there's no report.
     b) if there is data, there is one or more report.
2) Submit everything that was generated. If there was a problem with some of the data, `buildOneReport` should have returned an error.

Reply:
I don't think its no-op reports that should be an error, at this point in the code we have `priceUpdates` of nonzero length, so reports _must not_ be `nil`. All I'm saying is that if you submit a nonzero length price updates struct to `buildOneReport` and it returns `nil` this is a bug in `buildOneReport`. At this point it'd just be a sanity check that `report != nil` is the only case (i.e if the `if` condition on line 390 is not triggered, AND there is no error, there's something fatally wrong)